src/components/pages/search.js

```diff
@@ -35,14 +37,26 @@ function Search({ initialState }) {
 
   const [filteredDocs, setFilteredDocs] = useState([...sortedDocs]);
 
-  const [selectedTags, setSelectedTags] = useState([]);
+  const queryTagsToLower = new Set(decodedQuery.split(/\s/)
+    .filter(tag => tag.length > 3)
+    .map(tag => tag.toLowerCase()));
+
+  const { allTags, tagMap } = docs.reduce((acc, doc) => {
+    doc.tags.forEach(tag => {
+      acc.allTags.add(tag);
+      acc.tagMap[tag.toLowerCase()] = tag;
+    });
```

Comment:
To me this `forEach` in a `reduce` that processes 2 things at the same time looks kind of strange.
I think this construct might become simple/more clear by first getting all tags, then processing them.
Thinking of something like:

```
const allTags = docs.map(doc => doc.allTags).flat();
const allUniqueTags = new Set(allTags);
const tagMap = allTags.reduce(acc, tag => { acc[tag.toLowerCase()] = tag; return acc; }, {});
```

Reply:
Actually maybe there's no need to complicate things through this `tagMap` at all. Since it's used just once, I think you can just do this directly:

```
  const initialSelectedTags = allUniqueTags.filter(tag => queryTagsToLower.includes(tag.toLowerCase()));
```
I think this even saves you an iteration, just as trivia info :)

Also I guess `initialSelectedTags `would need to be in the original casing 🤔 The way you obtain them yields you the lowercased tags, but I guess the user needs to see the actual casing of the preselected tags, right? If it's so, then my suggestion covers this.

Reply:
The initialSelectedTags are in the casing done by the user, that was the point of that map. Have you tried the code?

Q: I think this even saves you an iteration, just as trivia info :). 
Let's assume that the number of query tags is n and the number of returned tags is m.
The tagMap is there to search in O(1) in the M entries + you have a O(n) because you search each tag once and another O(n) because you do the mapping and another O(n) for "filtering" the tags.
Putting it all together you have my algorithm:  n * O(1) + O(n) + O(n) = O(n) + O(n) + O(n) = 3 * O(n) = O(n)
With the other algo:
you have almost the same but the line:
 allUniqueTags.filter(tag => queryTagsToLower.includes(tag.toLowerCase()));
introduces an O(n * m) since you look for n tags into m elements = O(m*n)

Putting it all together O(m*n) > O(n) always :).

Reply:
TLDR thanks, I didn't take it as far, I said 'I think' 😄

Reply:
"The initialSelectedTags are in the casing exiting in the db which is the casing they were found in there, that was the point of that map. Have you tried the code?"
It's a suggestion, the ultime try has to be yours, the code maker. I think it should work.